### PR TITLE
Add cpu usage benchmark command `make cpu-benchmark` and website docs

### LINKED ISSUE #11077: Implement command to get cpu usage benchmark
/kind feature

Currently, minikube has several driver choices.
The drivers have each characteristics and they affect local machine's performance differently.

Of course, low CPU load and high performance are desirable, but it is important to know each characteristics and performance values ​​of each drivers.

In the past, Thomas took cpu usage benchmark using [cstat](https://github.com/tstromberg/cstat) on September 2020.
[Local Kubernetes - CPU Busy% overhead on macOS](https://docs.google.com/spreadsheets/d/12mBmRmNL-31fOQLjm4v0AvGc0iGwBrEHbQ7Uo_ysCnQ)
This benchmark compared minikube's drivers and similer products(Kind, K3d, Docker for Desktop).

Now if we'd like to take cpu usage benchmark, It takes long time and to summarize manually.
So, in this issue, I'll aim to implement the `make XXXX` command to do to reduce mannual costs.

### PR DESCRIPTION
**Pull request by govargo**
### What type of PR is this?
/kind feature

### What this PR does / why we need it:

This PR add cpu usage benchmark command `make cpu-benchmark` to minikube Makefile.
Once user take `make cpu-benchmark`, the script measures the cpu busy% using [cstat](https://github.com/tstromberg/cstat) while starting up minikube, docker, kind, k3d.
Measurment result will be summarized as a bar chart graph, and it will be placed as a png file under `site/static/image/cpuUsage`.

These scripts are for the benchmark of cpu usage, minikube vs kind vs k3d.   

 * `benchmark_local_k8s.sh`: take benchmark for cpu usage. This will take long to take place  
 * `update_summary.sh`: create one summary csv file of each drivers and products
 * `chart.go`: create bar chart graph as a png file

In macOS, we compare to minikube drivers(hyperkit, virtualbox, docker, docker with auto-pause) and kind, k3d, Docker for Mac Kubernetes
In Linux, we compare to minikube drivers(kvm2, virtualbox, docker, docker with auto-pause) and kind, k3d

### Which issue(s) this PR fixes:
Fix #11077

### Does this PR introduce a user-facing change?

No, this PR added `hack` scripts.
I added README about these scripts.

https://deploy-preview-11140--kubernetes-sigs-minikube.netlify.app/docs/benchmarks/cpuusage/

### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```
NONE
```

### CODE REVIEW

**Comment by afbjorklund** [comment]
If this benchmark is a shell script, why does it modify the go.mod and go.sum files ?

File: site/content/en/docs/benchmarks/cpuUsage/linux/_index.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+---
+title: "CPU Usage Benchmarks(Linux)"
+linkTitle: "CPU Usage Benchmarks(Linux)"
+weight: 1
+---
+
+![linux](/images/benchmarks/cpuUsage/linux.png)
```
Comment (afbjorklund) [inline site/content/en/docs/benchmarks/cpuUsage/linux/_index.md:7]: this image is missing

**Review by afbjorklund — CHANGES_REQUESTED** [review]

File: site/content/en/docs/benchmarks/cpuUsage/linux/_index.md (hunk 1 above)
Comment (govargo) [reply site/content/en/docs/benchmarks/cpuUsage/linux/_index.md:7]: Thank you for your pointing out!
I added the image.

**Comment by govargo** [comment]
Thank you afbjorklund.
There was not much description, so I updated this PR description.

And I added docker driver with auto-pause addon to benchmark, according to medya's advice.

**Comment by govargo** [comment]
https://deploy-preview-11140--kubernetes-sigs-minikube.netlify.app/docs/benchmarks/cpuusage/

**Review by medyagh — CHANGES_REQUESTED** [review]
KVM2 really taking more cpu than Docker on linux ? is it possible that you did not stop Docker service ? sudo systemctl stop docker ? and it was combining both idle docker + kvm2 ?

for auto-pause I would like to see a Separate Chart that clearly shows Long term effects of auto-pause

current chart is doing 
```
 start each local kubernetes tool
 2. deploy sample application(nginx deployment)
 3. wait 1 minute without anything
 4. measure No.3 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
 5. enable auto-pause addons(only if tool is minikube)
 6. wait 3 minute without anything
 7. measure No.6 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
```
but it should be doing :
```
1. start each local kubernetes tool
 2. deploy sample application(nginx deployment)
 3. wait 1 minute without anything
 4. measure No.3 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
 5. enable auto-pause addons(only if tool is minikube)
 6. wait 3 minute without anything
 7. measure No.6 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)

```

it should be Two new Charts

Auto-pause Chart -Paused
This chart shoudl be comparable to the First chart that Did NOT deploy anything.  
```
1. start each local kubernetes tool
2. Enable auto-pause)
3. wait 1 minute without anything (and verify it is paused) using minikube status
4. measure CPU usage with [cstat](https://github.com/tstromberg/cstat)

```

**Review by medyagh — CHANGES_REQUESTED** [review]
also I suggest we deploy nginix to all of the Experiments so we can compare them fairly

so a non-auto-pause should also have a ngnix on all tools

**Comment by govargo** [comment]
> KVM2 really taking more cpu than Docker on linux ? is it possible that you did not stop Docker service ? sudo systemctl stop docker ? and it was combining both idle docker + kvm2 ?

The script already stop docker engine by `sudo systemctl stop docker`.

https://github.com/kubernetes/minikube/pull/11140/files#diff-2e3d38e63fa9184219f976fdf6197997c5d691b065ed9ae6521eb9c9335f3003R269-R275

I measured on GCE with nested virtualization.
https://cloud.google.com/compute/docs/instances/enable-nested-virtualization-vm-instances
Maybe, it may affect the cpu usage by the overhead of nested virtualization.
I think virtualbox too.

I'll add notice about this environment caution to docs.

**Comment by govargo** [comment]
@medyagh 
Did you mention that should I separate to 2 charts(`Auto-pause Chart with deployment` and `Auto-pause Chart without deployment`)?
It looks like 'current chart is doing' is the same as your 'but it should be doing :'.
```
# current chart is doing
 start each local kubernetes tool
 2. deploy sample application(nginx deployment)
 3. wait 1 minute without anything
 4. measure No.3 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
 5. enable auto-pause addons(only if tool is minikube)
 6. wait 3 minute without anything
 7. measure No.6 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
```
```
# but it should be doing :
1. start each local kubernetes tool
 2. deploy sample application(nginx deployment)
 3. wait 1 minute without anything
 4. measure No.3 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
 5. enable auto-pause addons(only if tool is minikube)
 6. wait 3 minute without anything
 7. measure No.6 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
```

Please let me check your intention.
Thank you,

**Comment by govargo** [comment]
I noticed I took wrong way to measure for auto-pause.
Currently, I started measuring again. Measuring takes long time.

Tomorrow, I'll update comment and scripts and details and results.

**Comment by govargo** [comment]
Hi @medyagh

> I noticed I took wrong way to measure for auto-pause.
I improved how to measure auto-pause vs. non auto-pause.
Let me explain.

I noticed that I measured auto-pause cpu usage even though control plane status had been `Stopped` status, not `Paused` status.
Sorry, I overlook that auto-pause take 1 minute to become `Paused` status after auto-pause addon is enabled.
In other words, I measured by mixing the Stop state and the Pause state.
Of course, I should have measure only `Pause` state.

I improved how to measure auto-pause.
I added verification step which confirm if control plane is `Paused` status.
(You previously advised me this verification step, but I overlooked)

**Before**
 1. start each local kubernetes tool
 2. deploy sample application(nginx deployment) to each tool
 3. wait 1 minute without anything
 4. measure No.3 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)
 5. if tool is minikube, enable auto-pause addons(only if tool is minikube)
 6. if tool is minikube, wait 3 minute without anything
 7. if tool is minikube, measure No.6 idle CPU usage with [cstat](https://github.com/tstromberg/cstat)

**After**
 1. start each local kubernetes tool
 2. deploy sample application(nginx deployment) to each tool
 3. wait 1 minute without anything
 4. measure No.3 idle CPU usage with cstat
 5. if tool is minikube, enable auto-pause addon which pause control plane
 6. if tool is minikube, **wait 1 minute so that control plane will become Paused status(It takes 1 minute to become Pause status from Stopped status)**
 7. if tool is minikube, **verify if minikube control plane is paused**
 8. if tool is minikube, wait 3 minute without anything
 9. if tool is minikube, measure No.8 idle CPU usage with cstat

**Comment by govargo** [comment]
Here is the latest results for each OS.
These are very different from the old results, especially Linux.

## macOS
#### Old result(wrong measurement)
before-mac
#### Latest result(improved measurement)
latest-mac

## Linux
#### Old result(wrong measurement)
before-linux
#### Latest result(improved measurement)
latest-linux

**Comment by govargo** [comment]
I also added NOTE to README about GCE with nested virtualization.

https://github.com/kubernetes/minikube/pull/11140/files#diff-404c53b833a9c8717aea050a09899fd6415ba44d8ffabd20e4ccf097c2e7d7e4R40-R41

I think this PR is ready for review again.

**Review by medyagh — COMMENTED** [review]
Looks good to me just minor requests

**Review by medyagh — CHANGES_REQUESTED** [review]
1- can we remove the word Tools in the Chart Picture ?
2- for the Auto-Pause Chart, can we Change the color of Blue to Gray and Put in English
"Initial Start Before Pause" ?

3- Verify that for Auto-pause test the Ngnix is applied to kind as well ?


4- allso consider putting the commands for the user to Run this bechnmark themselves
like to run this benchmark run `.....`

**Comment by govargo** [comment]
> 1- can we remove the word Tools in the Chart Picture ?

Removed 'Tools' from chart.

> 2- for the Auto-Pause Chart, can we Change the color of Blue to Gray and Put in English "Initial Start Before Pause" ?

Changed color to gray and description 'Initial Start CPU usage Before Pause'.
https://deploy-preview-11140--kubernetes-sigs-minikube.netlify.app/docs/benchmarks/cpuusage/linux/
https://deploy-preview-11140--kubernetes-sigs-minikube.netlify.app/docs/benchmarks/cpuusage/macos/
gray-mac
gray-linux

If you would like to change color from gray to other, please let me know.

> 3- Verify that for Auto-pause test the Ngnix is applied to kind as well ?

Yes. I deployed nginx deployment to all tools(Of course, including minikube and kind, k3d).
I added the following description to our docs.
`We compare CPU usage after deploying sample application(nginx deployment) to all tools(including minikube and other tools).`
https://deploy-preview-11140--kubernetes-sigs-minikube.netlify.app/docs/benchmarks/cpuusage/linux/

> 4- allso consider putting the commands for the user to Run this bechnmark themselves
like to run this benchmark run .....

Anyone can run these benchmark by make command.
The commands are `make cpu-benchmark-idle` and `make cpu-benchmark-autopause`.
I already added description about these commands to README.md.
https://github.com/kubernetes/minikube/pull/11140/files#diff-f72cd7377ce0daf7fd7db606c6822f2ccf7f7d45614452f9e9c1b038a6dade11R25-R39

**Review by medyagh — APPROVED** [review]
Looks good to me @govargo  , Thank you for this PR and adding performance metrics, this will help our users to decide better

**Pull request merged**